xonsh/completer.py

```diff
@@ -257,3 +262,31 @@ def attr_complete(self, prefix, ctx):
             comp = prefix[:prelen-len(attr)] + rpl
             attrs.add(comp)
         return attrs
+
+    def _all_commands(self):
+        path = builtins.__xonsh_env__.get('PATH', None) or []
+        # did PATH change?
+        path_hash = hash(tuple(path))
+        cache_valid = path_hash == self._path_checksum
+        self._path_checksum = path_hash
+        # did aliases change?
+        al_hash = hash(tuple(builtins.aliases.keys()))
+        self._alias_checksum = al_hash
+        cache_valid = cache_valid and al_hash == self._alias_checksum
+        pm = self._path_mtime
+        # did the contents of any directory in PATH change?
+        for d in filter(os.path.isdir, path):
+            m = os.stat(d).st_mtime
+            if m > pm:
+                pm = m
+                cache_valid = False
```

Comment:
Right.  I went back and forth on whether to enter this loop if the cache had already been invalidated, and on whether to break here or not.  I think where I settled on those things was right.  This way, even if the cache is invalidated, we make sure we have up-to-date info in all of the cached things (path hash, alias hash, max path mtime).